### DISCUSSION TITLE
The proper way to do average from Oct to next Sep

### DISCUSSION THREAD

**Discussion by ShihengDuan** [xCDAT/xcdat#393] [2. Questions]
Hi, 

I'm trying to use xCDAT to do the water year average (or summation), which is from October in year ```n```, to September in year ```n+1```. This is basically a year average, and in ```xarray.Dataset.temporal.group_average``` I see we can customize seasons, but in my case, how should I specify the months from Oct to next Sep? Or should I use some other functions? 

Thanks!

**Comment by ShihengDuan** [comment]
What I do in ```xarray``` is to assign a new coordinate called ```WY```, and it equals to ```time.dt.year+(time.dt.month>9)``` (Oct, Nov and Dec are in the next water year). And then do a ```groupby``` and ```sum```. Finally remove the first and last years if they are incomplete. Is there any smarter way to do it with xcdat?

**Reply by pochedls** [reply]
Tagging @lee1043 as he helped design this API. I think there was discussion about custom seasons, but I have not used this feature. @ShihengDuan - did you see the documentation on [custom seasons](https://xcdat.readthedocs.io/en/stable/generated/xarray.Dataset.temporal.group_average.html)?

**Reply by lee1043** [reply]
I guess below might work, @ShihengDuan can you give a try?

```python
custom_seasons = [["Oct", "Nov", "Dec", "Jan", "Feb", "Mar", "Apr", "May", "Jun", "Jul", "Aug", "Sep"]]

ds_season_custom = ds.temporal.group_average(
    "ts",  # name of data array for variable
    "season",
    season_config={"custom_seasons": custom_seasons}
)
```

**Reply by ShihengDuan** [reply]
Yes, I noticed that we can specify the custom seasons. But in the documentation, it says 'Order of the months in each custom season does not matter'. So I don't think it is taking October this year to September of next year. I have a simple test case as the following:
```
data = np.ones((365*3))
data[365:365*2] = 2
data[365*2:365*3] = 3
start = np.datetime64('2009-01-01')
time_index = [start+np.timedelta64(i, 'D') for i in range(365*3)]
data = xa.DataArray(data, dims=['time'], coords={'time':time_index})
data.plot()
plt.show()
data.to_netcdf('test_data.nc')
```
![image](https://user-images.githubusercontent.com/26401994/202612135-b2d9ead1-897a-43f3-99a5-35460f2a5769.png)
That is, one value for each calendar year. So if we group by calendar year (Jan to Dec), it should be 1, 2, 3 for these three years, and some float values if it is taking the mean by water year (October to September). 
```
xc_data = xc.open_dataset('test_data.nc')
custom_seasons = [["Oct", "Nov", "Dec", "Jan", "Feb", "Mar", "Apr", "May", "Jun", "Jul", "Aug", "Sep"]]

xc_data_WY = xc_data.temporal.group_average(
    "__xarray_dataarray_variable__",  # name of data array for variable
    "season",
    season_config={"custom_seasons": custom_seasons}
)
plt.plot(xc_data_WY.__xarray_dataarray_variable__.data)
plt.show()
print(xc_data_WY.__xarray_dataarray_variable__.data)
```
And the result is here: 
<img width="436" alt="image" src="https://user-images.githubusercontent.com/26401994/202612422-f4858fe8-f2dc-4e05-bd35-c210f71a2373.png">

So I think this is still following the calendar year?

**Reply by pochedls** [reply]
Hi @ShihengDuan - I think you're right. It looks like you might have to write some custom code for this particular use case. I was going to suggest that this could be a potential new feature of xcdat where you can do this kind of thing more easily (if you were willing to try to take on adding this), but it does look like it could potentially be complicated to add in this functionality (it looks like DJF averaging is implemented by [shifting Decembers](https://github.com/xCDAT/xcdat/blob/main/xcdat/temporal.py#L1324) into the following year).

**Reply by ShihengDuan** [reply]
Thanks for pointing to the code of shifting Decembers. Then I think it is not straightforward to have a custom season that can handle the months in different years.

**Reply by arfriedman** [reply]
Hi @ShihengDuan @pochedls @lee1043 I wanted to echo interest in custom seasons which cross the year, and I was wondering if you've found any best-practices. For example, I often would like to calculate boreal winter averages from Dec-March.

Currently I create such averages by resampling over several months in xarray, but I would love to find a way to use xCDAT's ability to automatically account for different month lengths.

**Reply by jypeter** [reply]
@pochedls for custom seasons, we previously used this kind of thing `JJAS = cdutil.times.Seasons(‘JJAS’)` (see page 14 of `cdat_utilities.pdf` in https://github.com/xCDAT/xcdat/discussions/170)

I'm not sure if this is already in one of the existing time related Feature Request.

There is a good chance previous CDAT users need most of the features described in sections 1.1, 1.2 and 2.1 of `cdat_utilities.pdf`. You can probably ask legacy CDAT users at PCMDI what they think

**Reply by pochedls** [reply]
@arfriedman and @ShihengDuan – I can see that this would be very useful functionality and I think would be a good feature request. Would one of you be willing to add the feature request? Are either of you interested in putting together code to do this (I think @tomvothecoder might be able to provide guidance)?

@jypeter - there is a discussion thread on [New Features / Ideas](https://github.com/xCDAT/xcdat/discussions/271). I think that would be a good place to engage on prioritizing development going forward. The hope is that the broader community can help to maintain xCDAT and add new functionality, which will help keep xCDAT modern and up-to-date.

**Reply by tomvothecoder** [reply]
> @arfriedman and @ShihengDuan – I can see that this would be very useful functionality and I think would be a good feature request. Would one of you be willing to add the feature request? Are either of you interested in putting together code to do this (I think @tomvothecoder might be able to provide guidance)?

I agree that this sounds useful and is a good feature request. As @pochedls said, I encourage opening up a [Feature Request](https://github.com/xCDAT/xcdat/issues/new?assignees=&labels=Type%3A+Enhancement&template=feature_request.yml&title=%5BFeature%5D%3A+) and giving it a shot with a PR! Even submitting a draft PR with an xarray-based implementation can help lay the foundation for an xCDAT API implementation. @xCDAT/core-developers team can jump in the PR to help out too.

Here's xCDAT's [contributing guide](https://xcdat.readthedocs.io/en/latest/contributing.html) if there is interest.

**Reply by arfriedman** [reply]
Thank you @jypeter @tomvothecoder @pochedls.

I can work on putting together a feature request (it will be my first one for any project!).

**Reply by pochedls** [reply]
@arfriedman – thank you for being willing to put together the feature request - we appreciate the help. Let us know if you have any questions.
